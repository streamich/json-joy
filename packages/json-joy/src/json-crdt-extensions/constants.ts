export enum ExtensionId {
  mval = 0,
  cnt = 1,
  peritext = 2,
  quill = 3,
}

export enum ExtensionName {
  mval = ExtensionId.mval,
  cnt = ExtensionId.cnt,
  peritext = ExtensionId.peritext,
  quill = ExtensionId.quill,
}
