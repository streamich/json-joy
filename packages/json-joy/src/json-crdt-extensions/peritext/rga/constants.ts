export enum Anchor {
  Before = 0,
  After = 1,
}
