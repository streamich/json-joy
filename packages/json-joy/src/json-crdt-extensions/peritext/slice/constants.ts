/**
 * Specifies whether the start or the end of the cursor is the "anchor", e.g.
 * the end which does not move when user changes selection. The other
 * end is free to move, the moving end of the cursor is "focus". By default
 * "anchor" is usually the start of the cursor.
 */
export enum CursorAnchor {
  Start = 0,
  End = 1,
}

export const enum SliceTypeCon {
  // ---------------------------------------------------- block slices (0 to 64)
  p = 0, // <p>
  blockquote = 1, // <blockquote>
  codeblock = 2, // <pre><code>
  pre = 3, // <pre>
  ul = 4, // <ul>
  ol = 5, // <ol>
  tl = 6, // - [ ] Task list
  li = 7, // <li>
  h1 = 8, // <h1>
  h2 = 9, // <h2>
  h3 = 10, // <h3>
  h4 = 11, // <h4>
  h5 = 12, // <h5>
  h6 = 13, // <h6>
  title = 14, // <title>
  subtitle = 15, // <subtitle>
  br = 16, // <br>
  nl = 17, // \n
  hr = 18, // <hr>
  page = 19, // Page break
  aside = 20, // <aside>
  embed = 21, // <embed>, <iframe>, <object>, <video>, <audio>, etc.
  column = 22, // <div style="column-count: ..."> (represents 2 and 3 column layouts)
  contents = 23, // Table of contents
  table = 24, // <table>
  row = 25, // Table row
  cell = 26, // Table cell

  // TODO: rename to `cl` (collapsible list)?
  collapselist = 27, // Collapsible list - > List item

  collapse = 28, // Collapsible block
  note = 29, // Note block
  mathblock = 30, // <math> block
  div = 31,

  // ------------------------------------------------ inline slices (-64 to -1)
  Cursor = -1,
  RemoteCursor = -2,
  b = -3, // <b>
  i = -4, // <i>
  u = -5, // <u>
  s = -6, // <s>
  code = -7, // <code>
  mark = -8, // <mark>
  a = -9, // <a>
  comment = -10, // User comment attached to a slice
  del = -11, // <del>
  ins = -12, // <ins>
  sup = -13, // <sup>
  sub = -14, // <sub>
  math = -15, // <math> inline
  font = -16, // <span style="font-family: ...">
  col = -17, // <span style="color: ...">
  bg = -18, // <span style="background: ...">
  kbd = -19, // <kbd>
  spoiler = -20, // <span style="color: transparent; background: black">
  q = -21, // <q> (inline quote)
  cite = -22, // <cite> (inline citation)
  footnote = -23, // <sup> or <a> with href="#footnote-..." and title="Footnote ..."
  ref = -24, // <a> with href="#ref-..." and title="Reference ..." (Reference to some element in the document)
  iaside = -25, // Inline <aside>
  iembed = -26, // inline embed (any media, dropdown, Google Docs-like chips: date, person, file, etc.)
  bookmark = -27, // UI for creating a link to this slice
  overline = -28, // <span style="text-decoration: overline">
}

/**
 * All type name must be fully lowercase, as HTML custom element tag names must
 * be lowercase.
 */
export enum SliceTypeName {
  p = SliceTypeCon.p,
  blockquote = SliceTypeCon.blockquote,
  codeblock = SliceTypeCon.codeblock,
  pre = SliceTypeCon.pre,
  ul = SliceTypeCon.ul,
  ol = SliceTypeCon.ol,
  tl = SliceTypeCon.tl,
  li = SliceTypeCon.li,
  h1 = SliceTypeCon.h1,
  h2 = SliceTypeCon.h2,
  h3 = SliceTypeCon.h3,
  h4 = SliceTypeCon.h4,
  h5 = SliceTypeCon.h5,
  h6 = SliceTypeCon.h6,
  title = SliceTypeCon.title,
  subtitle = SliceTypeCon.subtitle,
  br = SliceTypeCon.br,
  nl = SliceTypeCon.nl,
  hr = SliceTypeCon.hr,
  page = SliceTypeCon.page,
  aside = SliceTypeCon.aside,
  embed = SliceTypeCon.embed,
  column = SliceTypeCon.column,
  contents = SliceTypeCon.contents,
  table = SliceTypeCon.table,
  row = SliceTypeCon.row,
  cell = SliceTypeCon.cell,
  collapselist = SliceTypeCon.collapselist,
  collapse = SliceTypeCon.collapse,
  note = SliceTypeCon.note,
  mathblock = SliceTypeCon.mathblock,
  div = SliceTypeCon.div,

  Cursor = SliceTypeCon.Cursor,
  RemoteCursor = SliceTypeCon.RemoteCursor,
  b = SliceTypeCon.b,
  i = SliceTypeCon.i,
  u = SliceTypeCon.u,
  s = SliceTypeCon.s,
  code = SliceTypeCon.code,
  mark = SliceTypeCon.mark,
  a = SliceTypeCon.a,
  comment = SliceTypeCon.comment,
  del = SliceTypeCon.del,
  ins = SliceTypeCon.ins,
  sup = SliceTypeCon.sup,
  sub = SliceTypeCon.sub,
  math = SliceTypeCon.math,
  font = SliceTypeCon.font,
  col = SliceTypeCon.col,
  bg = SliceTypeCon.bg,
  kbd = SliceTypeCon.kbd,
  spoiler = SliceTypeCon.spoiler,
  footnote = SliceTypeCon.footnote,
  ref = SliceTypeCon.ref,
  iaside = SliceTypeCon.iaside,
  iembed = SliceTypeCon.iembed,
  bookmark = SliceTypeCon.bookmark,
  overline = SliceTypeCon.overline,
}

/** Slice header octet (8 bits) masking specification. */
export enum SliceHeaderMask {
  /** The {@link Anchor} of the slice start {@link Point}.  */
  X1Anchor = 0b00000001,
  /** The {@link Anchor} of the slice end {@link Point}.  */
  X2Anchor = 0b00000010,
  /** Slice stacking behavior, one of {@link SliceStacking}. */
  Stacking = 0b00011100,
}

export enum SliceHeaderShift {
  X1Anchor = 0,
  X2Anchor = 0 + 1,
  Stacking = 0 + 1 + 1,
}

export enum SliceStacking {
  /**
   * The `Marker` slices are used to mark a block split position in the
   * document. For example, paragraph, heading, blockquote, etc. It separates
   * adjacent blocks and is used to determine the block type of the contents
   * following the marker, until the next marker is encountered.
   */
  Marker = 0b000,

  /**
   * The `Many` slices are inline formatting annotations, which allow one
   * or more annotations of the same type to apply to the same text. Slices with
   * stacking behavior `Many` are appended to the stack of attributes for a
   * specific slice type. With the most recent annotation on top.
   *
   * Slices with stacking behavior `Many` are used for inline formatting, like
   * for links, comments, etc. Where multiple annotations of the same type can
   * be applied to the same text.
   */
  Many = 0b001,

  /**
   * The slices with stacking behavior `One` are used for inline formatting
   * annotations, they overwrite the stack of attributes for a specific slice
   * type. This type of slice is used when only one annotation of a specific
   * type can be applied to the same text. For example, those could be used
   * for simple inline formatting, like bold, italic, etc.
   */
  One = 0b010,

  /**
   * The `Erase` slices are used to soft remove all annotations
   * (`Many` or `One`) of the same type which are applied to the same text
   * range. The erase slices soft remove only the annotations which were applied
   * before the erase slice, as determined by the logical clock (there could
   * be many layers of annotations applied and erased).
   *
   * Usually slices with stacking behavior `Erase` are used to reverse inline
   * exclusive (`One`) inline formatting, like bold, italic, etc.
   */
  Erase = 0b011,

  /**
   * Used to mark the user's cursor position in the document.
   *
   * @todo Consider removing this.
   */
  Cursor = 0b100,
}

export enum SliceStackingName {
  Marker = SliceStacking.Marker,
  Many = SliceStacking.Many,
  One = SliceStacking.One,
  Erase = SliceStacking.Erase,
  Cursor = SliceStacking.Cursor,
}

/**
 * Specifies `vec` offsets in the {@link SliceView}.
 */
export enum SliceTupleIndex {
  Header = 0,
  X1 = 1,
  X2 = 2,
  Type = 3,
  Data = 4,
}
