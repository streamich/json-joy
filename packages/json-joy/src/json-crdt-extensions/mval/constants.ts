import {ExtensionId, ExtensionName} from '../constants';

export const MNEMONIC = ExtensionName[ExtensionId.mval];
