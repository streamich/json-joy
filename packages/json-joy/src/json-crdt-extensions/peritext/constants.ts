import {type nodes, s} from '../../json-crdt-patch';
import {ExtensionId, ExtensionName} from '../constants';
import type {SliceSchema} from './slice/types';

export enum Chars {
  BlockSplitSentinel = '\n',
}

export enum Position {
  /**
   * Specifies the absolute start of the text, i.e. the position before the
   * first character. In model space it is defined as string ID and "after"
   * anchor.
   */
  AbsStart = -1,

  /**
   * Specifies the absolute end of the text, i.e. the position after the last
   * character. In model space it is defined as string ID and "before" anchor.
   */
  AbsEnd = 9007199254740991, // Number.MAX_SAFE_INTEGER
}

export const MNEMONIC = ExtensionName[ExtensionId.peritext];

export const SCHEMA = (text: string) =>
  s.vec<
    [
      /**
       * The text of the node. All rich-text textual data is stored in this node.
       */
      str: nodes.str<string>,
      /**
       * The slices of the node. All rich-text annotations are stored in this
       * node.
       */
      slices: nodes.arr<SliceSchema>,
    ]
  >(s.str<string>(text), s.arr<SliceSchema>([]));
