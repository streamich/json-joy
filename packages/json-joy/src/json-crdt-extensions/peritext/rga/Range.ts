import {Point} from './Point';
import {Anchor} from './constants';
import {updateNum} from '../../../json-hash/hash';
import type {ITimestampStruct} from '../../../json-crdt-patch/clock';
import type {Printable} from 'tree-dump/lib/types';
import type {AbstractRga, Chunk} from '../../../json-crdt/nodes/rga';
import type {Stateful} from '../types';

/**
 * A range is a pair of points that represent a selection in the text. A range
 * can be collapsed to a single point, then it is called a *marker*
 * (if it is stored in the text), or *caret* (if it is a cursor position).
 */
export class Range<T = string> implements Pick<Stateful, 'refresh'>, Printable {
  /**
   * Creates a range from two points. The points are ordered so that the
   * start point is before or equal to the end point.
   *
   * @param rga Peritext context.
   * @param p1 Some point.
   * @param p2 Another point.
   * @returns Range with points in correct order.
   */
  public static from<T = string>(rga: AbstractRga<T>, p1: Point<T>, p2: Point<T>): Range<T> {
    return p1.cmpSpatial(p2) > 0 ? new Range(rga, p2, p1) : new Range(rga, p1, p2);
  }

  /**
   * A convenience method for creating a range from a view position and a length.
   * The `start` argument specifies the position between characters, where
   * the range should start. The `size` argument specifies the number of
   * characters in the range. If `size` is zero or not specified, the range
   * will be collapsed to a single point.
   *
   * When the range is collapsed, the anchor position is set to "after" the
   * character. When the range is expanded, the anchor positions are set to
   * "before" for the start point and "after" for the end point.
   *
   * The `size` argument can be negative, in which case the range is selected
   * backwards.
   *
   * @param rga Peritext context.
   * @param start Position in the text between characters.
   * @param size Length of the range. Can be negative, in which case the range
   *             is selected backwards.
   * @returns A range from the given position with the given length.
   */
  public static at<T = string>(rga: AbstractRga<T>, start: number, size: number = 0): Range<T> {
    const length = rga.length();
    if (!size) {
      if (start > length) start = length;
      const startId = !start ? rga.id : rga.find(start - 1) || rga.id;
      const point = new Point(rga, startId, Anchor.After);
      return new Range<T>(rga, point, point.clone());
    }
    if (size < 0) {
      size = -size;
      start -= size;
    }
    if (start < 0) {
      size += start;
      start = 0;
      if (size < 0) return Range.at(rga, start, 0);
    }
    if (start >= length) {
      start = length;
      size = 0;
    }
    if (start + size > length) size = length - start;
    const startId = rga.find(start) || rga.id;
    const endId = rga.find(start + size - 1) || startId;
    const startEndpoint = new Point(rga, startId, Anchor.Before);
    const endEndpoint = new Point(rga, endId, Anchor.After);
    return new Range(rga, startEndpoint, endEndpoint);
  }

  /**
   * @param rga Peritext context.
   * @param start Start point of the range, must be before or equal to end.
   * @param end End point of the range, must be after or equal to start.
   */
  constructor(
    protected readonly rga: AbstractRga<T>,
    public start: Point<T>,
    public end: Point<T>,
  ) {}

  /**
   * Clones the range.
   *
   * @returns A new range with the same start and end points.
   */
  public range(): Range<T> {
    return new Range(this.rga, this.start.clone(), this.end.clone());
  }

  public cmp(range: Range<T>): -1 | 0 | 1 {
    return this.start.cmp(range.start) || this.end.cmp(range.end);
  }

  public cmpSpatial(range: Range<T>): number {
    return this.start.cmpSpatial(range.start) || this.end.cmpSpatial(range.end);
  }

  /**
   * Determines if the range is collapsed to a single point. Handles special
   * cases where the range is collapsed, but the points are not equal, for
   * example, when the characters between the points are invisible.
   *
   * @returns True if the range is collapsed to a single point.
   */
  public isCollapsed(): boolean {
    const {start, end} = this;
    if (start.cmpSpatial(end) === 0) return true;
    const start2 = start.clone();
    const end2 = end.clone();
    start2.refAfter();
    end2.refAfter();
    return start2.cmp(end2) === 0;
  }

  public contains(range: Range<T>): boolean {
    return this.start.cmpSpatial(range.start) <= 0 && this.end.cmpSpatial(range.end) >= 0;
  }

  public containsPoint(point: Point<T>): boolean {
    return this.start.cmpSpatial(point) <= 0 && this.end.cmpSpatial(point) >= 0;
  }

  // ---------------------------------------------------------------- mutations

  public set(start: Point<T>, end: Point<T> = start): void {
    this.start = start;
    this.end = end === start ? end.clone() : end;
  }

  public setRange(range: Range<T>): void {
    this.set(range.start, range.end);
  }

  public setAt(start: number, length: number = 0): void {
    const range = Range.at<T>(this.rga, start, length);
    this.setRange(range);
  }

  public setAfter(id: ITimestampStruct): void {
    const point = new Point(this.rga, id, Anchor.After);
    this.set(point);
  }

  /**
   * Collapse the range to the start point and sets the anchor position to be
   * "after" the character.
   */
  public collapseToStart(): void {
    const start = this.start.clone();
    start.refAfter();
    const end = start.clone();
    this.set(start, end);
  }

  /**
   * Collapse the range to the end point and sets the anchor position to be
   * "before" the character.
   */
  public collapseToEnd(): void {
    const end = this.end.clone();
    end.refAfter();
    const start = this.end.clone();
    this.set(start, end);
  }

  /**
   * Expand range left and right to contain all invisible space: (1) tombstones,
   * (2) anchors of non-deleted adjacent chunks.
   */
  public expand(): void {
    this.start.refAfter();
    this.end.refBefore();
  }

  /**
   * The reverse of {@link expand}. Shrink the range's start and end points to
   * still contain the same visible text, but narrow the range in the CRDT-space
   * as much as possible.
   */
  public shrink(): void {
    this.start.refBefore();
    this.end.refAfter();
  }

  // -------------------------------------------------- View coordinate methods

  /**
   * Returns the range in the view coordinates as a position and length.
   *
   * @returns The range as a view position and length.
   */
  public view(): [start: number, size: number] {
    const start = this.start.viewPos();
    const end = this.end.viewPos();
    return [start, end - start];
  }

  /**
   * @returns The length of the range in view coordinates.
   */
  public length(): number {
    return this.end.viewPos() - this.start.viewPos();
  }

  /**
   * Returns plain text view of the range. Concatenates all text chunks in the
   * range ignoring tombstones and returns the result.
   *
   * @returns The text content of the range.
   */
  public text(): string {
    const isCaret = this.isCollapsed();
    if (isCaret) return '';
    const {start, end} = this;
    const rga = this.rga;
    const startId = start.anchor === Anchor.Before ? start.id : start.nextId();
    const endId = end.anchor === Anchor.After ? end.id : end.prevId();
    if (!startId || !endId) return '';
    let result = '';
    rga.range0(start.chunk(), startId, endId, (chunk: Chunk<T>, from: number, length: number) => {
      const data = chunk.data as string;
      if (data) result += data.slice(from, from + length);
    });
    return result;
  }

  // ----------------------------------------------------------------- Stateful

  public refresh(): number {
    let state = this.start.refresh();
    state = updateNum(state, this.end.refresh());
    return state;
  }

  // ---------------------------------------------------------------- Printable

  public toStringName(): string {
    return 'Range';
  }

  public toString(tab: string = '', lite: boolean = true): string {
    const name = this.toStringName();
    const start = this.start.toString(tab, lite);
    const end = this.end.toString(tab, lite);
    let text = this.text();
    if (text.length > 16) text = text.slice(0, 16) + '...';
    return `${name} ${JSON.stringify(text)} ${start} â†” ${end}`;
  }
}
