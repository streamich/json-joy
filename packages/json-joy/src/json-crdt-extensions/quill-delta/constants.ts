import {ExtensionId, ExtensionName} from '../constants';

export enum QuillConst {
  EmbedChar = '\n',
  EmbedSliceType = 0,
}

export const MNEMONIC = ExtensionName[ExtensionId.quill];
