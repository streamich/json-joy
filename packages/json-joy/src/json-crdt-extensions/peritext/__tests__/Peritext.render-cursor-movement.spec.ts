import {Anchor} from '../rga/constants';
import {render} from './render';
import {
  type Kit,
  setupAlphabetChunkSplitKit,
  setupAlphabetKit,
  setupAlphabetWithDeletesKit,
  setupAlphabetWithTwoChunksKit,
  setupAlphabetWrittenInReverse,
  setupAlphabetWrittenInReverseWithDeletes,
} from './setup';

const runInlineSlicesTests = (desc: string, getKit: () => Kit) => {
  const setup = () => {
    const kit = getKit();
    const view = () => {
      kit.peritext.refresh();
      return render(kit.peritext.blocks.root, '', false);
    };
    // console.log(peritext.str + '');
    return {...kit, view};
  };

  describe(desc, () => {
    test('can move cursor forward - starting from middle', () => {
      const {editor, peritext, view} = setup();
      editor.cursor.setAt(1);
      peritext.refresh();
      expect(view()).toMatchInlineSnapshot(`
"<>
  <0>
    "a" {  }
    "bcdefghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`);
      editor.cursor.move(1);
      peritext.refresh();
      expect(view()).toMatchInlineSnapshot(`
"<>
  <0>
    "ab" {  }
    "cdefghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`);
      editor.cursor.move(2);
      peritext.refresh();
      expect(view()).toMatchInlineSnapshot(`
"<>
  <0>
    "abcd" {  }
    "efghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`);
    });

    test('can move cursor forward - starting the beginning of the string', () => {
      const {editor, peritext, view} = setup();
      editor.cursor.setAt(0);
      peritext.refresh();
      // console.log(view());
      expect(view()).toMatchInlineSnapshot(`
"<>
  <0>
    "abcdefghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`);
      editor.cursor.move(1);
      peritext.refresh();
      // console.log(view());
      // console.log(peritext + '');
      expect(view()).toMatchInlineSnapshot(`
"<>
  <0>
    "a" {  }
    "bcdefghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`);
      editor.cursor.move(2);
      peritext.refresh();
      // console.log(view());
      expect(view()).toMatchInlineSnapshot(`
"<>
  <0>
    "abc" {  }
    "defghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`);
    });

    test('can move cursor backward - starting from middle', () => {
      const {editor, peritext} = setup();
      editor.cursor.setAt(5);
      peritext.refresh();
      expect(editor.cursor.start.anchor).toBe(Anchor.After);
      expect(editor.cursor.start.pos()).toEqual(4);
      editor.cursor.move(-1);
      peritext.refresh();
      expect(editor.cursor.start.pos()).toEqual(3);
      editor.cursor.move(-2);
      peritext.refresh();
      expect(editor.cursor.start.pos()).toEqual(1);
      editor.cursor.move(-1);
      peritext.refresh();
      expect(editor.cursor.start.pos()).toEqual(0);
    });

    test('can move cursor backward - starting from end of string', () => {
      const {editor, peritext} = setup();
      editor.cursor.setAt(26);
      peritext.refresh();
      expect(editor.cursor.start.anchor).toBe(Anchor.After);
      expect(editor.cursor.start.pos()).toEqual(25);
      editor.cursor.move(-1);
      peritext.refresh();
      expect(editor.cursor.start.pos()).toEqual(24);
      editor.cursor.move(-2);
      peritext.refresh();
      expect(editor.cursor.start.pos()).toEqual(22);
    });

    test('can move cursor backward - until the beginning of the string', () => {
      const {editor, peritext} = setup();
      editor.cursor.setAt(5);
      peritext.refresh();
      expect(editor.cursor.start.anchor).toBe(Anchor.After);
      expect(editor.cursor.start.pos()).toEqual(4);
      editor.cursor.move(-2);
      peritext.refresh();
      expect(editor.cursor.start.pos()).toEqual(2);
      editor.cursor.move(-3);
      peritext.refresh();
      expect(editor.cursor.start.pos()).toEqual(-1);
    });

    test('can move cursor backward - more than there is space', () => {
      const {editor, peritext} = setup();
      editor.cursor.setAt(5);
      peritext.refresh();
      expect(editor.cursor.start.anchor).toBe(Anchor.After);
      expect(editor.cursor.start.pos()).toEqual(4);
      editor.cursor.move(-111111);
      peritext.refresh();
      expect(editor.cursor.start.pos()).toEqual(-1);
    });
  });
};

runInlineSlicesTests('single text chunk', setupAlphabetKit);
runInlineSlicesTests('two chunks', setupAlphabetWithTwoChunksKit);
runInlineSlicesTests('with chunk split', setupAlphabetChunkSplitKit);
runInlineSlicesTests('with deletes', setupAlphabetWithDeletesKit);
runInlineSlicesTests('written in reverse', setupAlphabetWrittenInReverse);
runInlineSlicesTests('written in reverse with deletes', setupAlphabetWrittenInReverseWithDeletes);
