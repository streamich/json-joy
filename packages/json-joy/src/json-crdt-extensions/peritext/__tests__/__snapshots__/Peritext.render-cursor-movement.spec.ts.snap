// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`single text chunk can move cursor forward - starting from middle 1`] = `
"<>
  <0>
    "a" {}
    "bcdefghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`single text chunk can move cursor forward - starting from middle 2`] = `
"<>
  <0>
    "ab" {}
    "cdefghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`single text chunk can move cursor forward - starting from middle 3`] = `
"<>
  <0>
    "abcd" {}
    "efghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`single text chunk can move cursor forward - starting the beginning of the string 1`] = `
"<>
  <0>
    "abcdefghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`single text chunk can move cursor forward - starting the beginning of the string 2`] = `
"<>
  <0>
    "a" {}
    "bcdefghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`single text chunk can move cursor forward - starting the beginning of the string 3`] = `
"<>
  <0>
    "abc" {}
    "defghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`two chunks can move cursor forward - starting from middle 1`] = `
"<>
  <0>
    "a" {}
    "bcdefghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`two chunks can move cursor forward - starting from middle 2`] = `
"<>
  <0>
    "ab" {}
    "cdefghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`two chunks can move cursor forward - starting from middle 3`] = `
"<>
  <0>
    "abcd" {}
    "efghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`two chunks can move cursor forward - starting the beginning of the string 1`] = `
"<>
  <0>
    "abcdefghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`two chunks can move cursor forward - starting the beginning of the string 2`] = `
"<>
  <0>
    "a" {}
    "bcdefghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`two chunks can move cursor forward - starting the beginning of the string 3`] = `
"<>
  <0>
    "abc" {}
    "defghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`with chunk split can move cursor forward - starting from middle 1`] = `
"<>
  <0>
    "a" {}
    "bcdefghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`with chunk split can move cursor forward - starting from middle 2`] = `
"<>
  <0>
    "ab" {}
    "cdefghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`with chunk split can move cursor forward - starting from middle 3`] = `
"<>
  <0>
    "abcd" {}
    "efghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`with chunk split can move cursor forward - starting the beginning of the string 1`] = `
"<>
  <0>
    "abcdefghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`with chunk split can move cursor forward - starting the beginning of the string 2`] = `
"<>
  <0>
    "a" {}
    "bcdefghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`with chunk split can move cursor forward - starting the beginning of the string 3`] = `
"<>
  <0>
    "abc" {}
    "defghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`with deletes can move cursor forward - starting from middle 1`] = `
"<>
  <0>
    "a" {}
    "bcdefghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`with deletes can move cursor forward - starting from middle 2`] = `
"<>
  <0>
    "ab" {}
    "cdefghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`with deletes can move cursor forward - starting from middle 3`] = `
"<>
  <0>
    "abcd" {}
    "efghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`with deletes can move cursor forward - starting the beginning of the string 1`] = `
"<>
  <0>
    "abcdefghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`with deletes can move cursor forward - starting the beginning of the string 2`] = `
"<>
  <0>
    "a" {}
    "bcdefghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`with deletes can move cursor forward - starting the beginning of the string 3`] = `
"<>
  <0>
    "abc" {}
    "defghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`written in reverse can move cursor forward - starting from middle 1`] = `
"<>
  <0>
    "a" {}
    "bcdefghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`written in reverse can move cursor forward - starting from middle 2`] = `
"<>
  <0>
    "ab" {}
    "cdefghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`written in reverse can move cursor forward - starting from middle 3`] = `
"<>
  <0>
    "abcd" {}
    "efghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`written in reverse can move cursor forward - starting the beginning of the string 1`] = `
"<>
  <0>
    "abcdefghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`written in reverse can move cursor forward - starting the beginning of the string 2`] = `
"<>
  <0>
    "a" {}
    "bcdefghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`written in reverse can move cursor forward - starting the beginning of the string 3`] = `
"<>
  <0>
    "abc" {}
    "defghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`written in reverse with deletes can move cursor forward - starting from middle 1`] = `
"<>
  <0>
    "a" {}
    "bcdefghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`written in reverse with deletes can move cursor forward - starting from middle 2`] = `
"<>
  <0>
    "ab" {}
    "cdefghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`written in reverse with deletes can move cursor forward - starting from middle 3`] = `
"<>
  <0>
    "abcd" {}
    "efghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`written in reverse with deletes can move cursor forward - starting the beginning of the string 1`] = `
"<>
  <0>
    "abcdefghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`written in reverse with deletes can move cursor forward - starting the beginning of the string 2`] = `
"<>
  <0>
    "a" {}
    "bcdefghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;

exports[`written in reverse with deletes can move cursor forward - starting the beginning of the string 3`] = `
"<>
  <0>
    "abc" {}
    "defghijklmnopqrstuvwxyz" { -1 = [ !u ] }
"
`;
